lib/redis_memo/memoize_method.rb

```diff
@@ -81,14 +81,53 @@ def self.extract_dependencies(ref, *method_args, &depends_on)
 
   def self.get_or_extract_dependencies(ref, *method_args, &depends_on)
     if RedisMemo::Cache.local_dependency_cache
-      RedisMemo::Cache.local_dependency_cache[ref] ||= {}
-      RedisMemo::Cache.local_dependency_cache[ref][depends_on] ||= {}
-      RedisMemo::Cache.local_dependency_cache[ref][depends_on][method_args] ||= extract_dependencies(ref, *method_args, &depends_on)
+      RedisMemo::Cache.local_dependency_cache[ref.class] ||= {}
+      RedisMemo::Cache.local_dependency_cache[ref.class][depends_on] ||= {}
+      mapped_args = map_depends_on_method_args(depends_on, ref, method_args)
+      RedisMemo::Cache.local_dependency_cache[ref.class][depends_on][mapped_args] ||= extract_dependencies(ref, *method_args, &depends_on)
     else
       extract_dependencies(ref, *method_args, &depends_on)
     end
   end
 
+  # We only look at method parameters that the dependency block cares about in order to define its dependent
+  # memos, following the convention that nil or :_ means we don't care about the parameter.
+  # Example:
+  # ```
+  #    def method(param1, param2)
+  #    end
+  #
+  #    memoize_method :method do |_, _, param2|`
+  #      depends_on RedisMemo::Memoizable.new(param2: param2)
+  #    end
+  # ```
+  #  `map_depends_on_method_args(depends_on, ref, [1, 2])` returns { param2 : 2 }
+  def self.map_depends_on_method_args(depends_on, ref, args)
+    mapped_args = {}
+    unless depends_on.parameters.empty? or args.empty?
+      depends_on_args = [ref] + args
+      depends_on.parameters.each_with_index do |param, i|
+        unless param.size != 2 || param[1] == :_
+          # If it's a single splat,
+          if param[0] == :rest
+            # take the rest of the arguments if this is the last parameter
+            if i == depends_on.parameters.size - 1
+              mapped_args[param[1]] = depends_on_args[i..-1]
+            # find the last element that isn't a hash, and replace those elements with a single array
+            else
+              splat_args = depends_on_args[i..-1].select { |x| !x.is_a?(Hash) }
```

Comment:
Hmmm, can you explain a little what are we doing here?

It's odd to me that we're selecting non_hash args from depends_on_args for each parameter (nested loops)

Is the `splat_args`  supposed to be `keyword_args`? If that's true, the keyword args hash should be:
```
remove_positional_args!(args)
if args.last.is_a?(Hash)
  args.last
else
  {}
end
```
separately, the logic here is dense. Let's add separate methods to keep this readable?

Reply:
I was adding a test case for the following scenario, when there's a splat, required keyword arguments vs the optional keyword arguments (**). For example:
```ruby
memoize_method :test do |_, a, *args,  b:, **kwargs| end

test(1, 2, 3, 4, b: 5, c: 6)
```
the mapped arguments should be
```ruby
mapped_args = {:a => 1, :args=> [2, 3, 4], :b => 5, :kwargs => {:c => 6} }
```
The tests aren't clearly testing for whether we do this mapping right though. And also I think there's some problems with the current logic, so will fix

Reply:
I tried to simplify the logic a bit so its a bit easier to follow, and also tried to test that the `exclude_anonymous_args` method is populating the local cache with what we expect. Hopefully it is less convoluted now